File: src/main/scala/com/github/ozancicek/artan/ml/SparkFunctions.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.ozancicek.artan.ml
import org.apache.spark.sql.functions.{array, col, lit, randn, udf}
import org.apache.spark.ml.linalg.{DenseMatrix, DenseVector, Matrices, Matrix, Vector, Vectors}
import org.apache.spark.ml.{BLAS, LAPACK}
import org.apache.spark.sql.Column
import org.apache.spark.sql.expressions.UserDefinedFunction


object SparkFunctions {

  private def cholesky(in: Matrix): Matrix = {
    val result = in.toDense.copy
    LAPACK.dpotrf(result)
    result
  }

  private def randnVector(size: Int, seed: Long): Column = {
    arrayToVector(array((0 until size).map(i => randn(seed + i)): _*))
  }

  private def choleskyUDF: UserDefinedFunction = udf((in: Matrix) => cholesky(in))


  /**
   * UDF converting array of doubles to vector
   *
   * @return UserDefinedFunction
   */
  def arrayToVector: UserDefinedFunction = udf { arr: Seq[Double] =>
    Option(arr).map(in => Vectors.dense(in.toArray))
  }

  /**
   * UDF converting vector to array
   *
   * @return UserDefinedFunction
   */
  def vectorToArray: UserDefinedFunction = udf { vec: Vector =>
    Option(vec).map(in => in.toArray.toSeq)
  }

  /**
   * UDF for creating DenseMatrix from column major array of doubles
   *
   * @return UserDefinedFunction
   */
  def arrayToMatrix: UserDefinedFunction = udf { (numRows: Int, numCols: Int, arr: Seq[Double]) =>
    new DenseMatrix(numRows, numCols, arr.toArray)
  }

  /**
   * UDF for creating vector of zeros
   *
   * @return UserDefinedFunction
   */
  def zerosVector: UserDefinedFunction = udf { size: Int => Vectors.zeros(size)}

  /**
   * UDF for creating vector of ones
   *
   * @return UserDefinedFunction
   */
  def onesVector: UserDefinedFunction = udf { size: Int => new DenseVector(Array.fill(size) {1.0})}

  /**
   * UDF for dot product of two vectors
   *
   * @return UserDefinedFunction
   */
  def dotVector: UserDefinedFunction = udf { (x: Vector, y: Vector) =>
    (Option(x), Option(y)) match {
      case (Some(_), Some(_)) => Some(BLAS.dot(x, y))
      case _ => None
    }
  }

  /**
   * UDF for scaling vector with a scalar constant
   *
   * @return UserDefinedFunction
   */
  def scalVector: UserDefinedFunction = udf { (alpha: Double, x: Vector) =>
    (Option(alpha), Option(x)) match {
      case (Some(_), Some(_)) => {
        val result = x.copy
        BLAS.scal(alpha, result)
        Some(result)
      }
      case _ => None
    }
  }

  /**
   * UDF for axpy operation on vectors, alpha*x + y
   *
   * @return UserDefinedFunction
   */
  def axpyVector: UserDefinedFunction = udf { (alpha: Double, x: Vector, y: Vector) =>
    (Option(alpha), Option(x), Option(y)) match {
      case (Some(_), Some(_), Some(_)) => {
        val result = y.copy
        BLAS.axpy(alpha, x, result)
        Some(result)
      }
      case _ => None
    }
  }

  /**
   * UDF for creating identity matrix
   *
   * @return UserDefinedFunction
   */
  def eyeMatrix: UserDefinedFunction = udf { size: Int => Matrices.eye(size)}

  /**
   * UDF for creating zeros matrix
   *
   * @return UserDefinedFunction
   */
  def zerosMatrix: UserDefinedFunction = udf { (numRows: Int, numCols: Int) => Matrices.ones(numRows, numCols)}

  /**
   * UDF for creating diagonal matrix from vector
   *
   * @return UserDefinedFunction
   */
  def diagMatrix: UserDefinedFunction = udf { diag: Vector => Matrices.diag(diag)}

  /**
   * UDF for outer product of two vectors, a*x*y.T
   *
   * @return UserDefinedFunction
   */
  def outerProduct: UserDefinedFunction = udf { (alpha: Double, x: Vector, y: Vector) =>
    val result = Matrices.zeros(x.size, y.size).toDense
    BLAS.dger(alpha, x.toDense, y.toDense, result)
  }

  /**
   * UDF for scaling vector of unit normal samples to multivariate gaussian with mean and covariance specified with its
   * root. root covariance should be lower triangle, and it can be obtained with cholesky decomposition
   * of full covariance matrix.
   *
   * @return UserDefinedFunction
   */
  def scaleToMultiGaussian: UserDefinedFunction = udf { (mean: Vector, covRoot: Matrix, normal: Vector) =>
    val scaled = mean.toDense.copy
    BLAS.gemv(1.0, covRoot, normal, 1.0, scaled)
    scaled
  }

  /**
   * Sample from multivariate gaussian with literal distribution parameters
   *
   * @param mean mean vector
   * @param covariance covariance matrix
   * @param seed seed for random number generation
   * @return Column
   */
  def randMultiGaussian(mean: Vector, covariance: Matrix, seed: Long = 0): Column = {
    val root = cholesky(covariance)
    scaleToMultiGaussian(lit(mean), lit(root), randnVector(mean.size, seed))
  }

  /**
   * Sample from multivariate gaussian with column distribution parameters
   *
   * @param meanCol mean vector column
   * @param covCol covariance matrix column
   * @param size size of the mean vector
   * @param seed seed for random number generation
   * @return Column
   */
  def randMultiGaussianWithCol(meanCol: String, covCol: String, size: Int, seed: Long = 0): Column = {
    scaleToMultiGaussian(col(meanCol), choleskyUDF(col(covCol)), randnVector(size, seed))
  }

}

File: python/artan/spark_functions.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at

#  http://www.apache.org/licenses/LICENSE-2.0

#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#


from pyspark.sql.column import (
    Column, _to_java_column, _to_seq
)
from pyspark import SparkContext
import pyspark.sql.functions as F
import numpy as np


def _sc():
    return SparkContext._active_spark_context


def _spark_functions():
    return _sc()._jvm.com.github.ozancicek.artan.ml.SparkFunctions


def _function_factory(cols, func):
    return Column(func.apply(_to_seq(_sc(), cols, _to_java_column)))


def arrayToVector(arrayCol):
    """
    Converts array of doubles column to dense vector column

    :param arrayCol: Array of doubles column, either string column name or column expression
    :return: DenseVector column
    """
    return _function_factory([arrayCol], _spark_functions().arrayToVector())


def vectorToArray(vectorCol):
    """
    Converts vector column to array of doubles

    :param vectorCol: Vector column, either string column name or column expression
    :return: Array of doubles column
    """
    return _function_factory([vectorCol], _spark_functions().vectorToArray())


def arrayToMatrix(numRowsCol, numColsCol, arrayCol):
    """
    Creates dense matrix from array of doubles column

    :param numRowsCol: Number of rows column, either integer or column expression
    :param numColsCol: Number of cols column, either integer or column expression
    :param arrayCol: Array of doubles column expression, should be column major ordered
    :return: DenseMatrix column
    """
    numRows = F.lit(numRowsCol) if isinstance(numRowsCol, int) else numRowsCol
    numCols = F.lit(numColsCol) if isinstance(numColsCol, int) else numColsCol
    return _function_factory([numRows, numCols, arrayCol], _spark_functions().arrayToMatrix())


def matrixToArray(matrixCol):
    """
    Converts matrix column to array of doubles

    :param matrixCol: Matrix column, either string column name or column expression
    :return: struct of (numRows, numCols, values) where values are column major values of matrix
    """
    return _function_factory([matrixCol], _spark_functions().matrixToArray())


def zerosVector(sizeCol):
    """
    Creates vector of zeros column

    :param sizeCol: Size of the vector, either integer or column expression
    :return: Vector column
    """
    size = F.lit(sizeCol) if isinstance(sizeCol, int) else sizeCol
    return _function_factory([size], _spark_functions().zerosVector())


def onesVector(sizeCol):
    """
    Creates vector of ones column

    :param sizeCol: Size of the vector, either integer or column expression
    :return: Vector column
    """
    size = F.lit(sizeCol) if isinstance(sizeCol, int) else sizeCol
    return _function_factory([size], _spark_functions().onesVector())


def dotVector(leftVectorCol, rightVectorCol):
    """
    Returns the dot product of two vector columns

    :param leftVectorCol: Vector column
    :param rightVectorCol: Vector column
    :return: Double column
    """
    return _function_factory([leftVectorCol, rightVectorCol], _spark_functions().dotVector())


def scalVector(scaleCol, vectorCol):
    """
    Scales the vector column with a scalar constant

    :param scaleCol: scale constant, either numeric or column expression
    :param vectorCol: Vector column
    :return: Vector column
    """
    scale = F.lit(float(scaleCol)) if isinstance(scaleCol, (int, float)) else scaleCol
    return _function_factory([scale, vectorCol], _spark_functions().scalVector())


def axpyVector(alphaCol, xVectorCol, yVectorCol):
    """
    Axpy operation on vectors, alpha*x +y where alpha is scalar, x and y vectors

    :param alphaCol: scalar which scales x, either numeric or column expression
    :param xVectorCol: Vector column corresponding to x
    :param yVectorCol: Vector column corresponding to y
    :return: Vector column
    """
    alpha = F.lit(float(alphaCol)) if isinstance(alphaCol, (int, float)) else alphaCol
    return _function_factory([alpha, xVectorCol, yVectorCol], _spark_functions().axpyVector())


def eyeMatrix(sizeCol):
    """
    Creates identity matrix column

    :param sizeCol: Size of the matrix, either integer or column expression
    :return: DenseMatrix column
    """
    size = F.lit(sizeCol) if isinstance(sizeCol, int) else sizeCol
    return _function_factory([size], _spark_functions().eyeMatrix())


def zerosMatrix(numRowsCol, numColsCol):
    """
    Creates zeros matrix column

    :param numRowsCol: Number of rows column, either integer or column expression
    :param numColsCol: Number of cols column, either integer or column expression
    :return: Matrix column
    """
    numRows = F.lit(numRowsCol) if isinstance(numRowsCol, int) else numRowsCol
    numCols = F.lit(numColsCol) if isinstance(numColsCol, int) else numColsCol
    return _function_factory([numRows, numCols], _spark_functions().zerosMatrix())


def diagMatrix(diagVectorCol):
    """
    Creates a diag matrix from vector

    :param diagVectorCol: Vector column corresponding to diagonal
    :return: Matrix column
    """
    return _function_factory([diagVectorCol], _spark_functions().diagMatrix())


def multiplyMatrix(leftCol, rightCol):
    """
    Matrix multiplication

    :param leftCol: Matrix column
    :param rightCol: Matrix column
    :return: Matrix column for the result of matrix multiplication
    """
    return _function_factory([leftCol, rightCol], _spark_functions().multiplyMatrix())


def multiplyMatrixVector(matCol, vecCol):
    """
    Multiply matrix with a vector

    :param matCol: Matrix column
    :param vecCol: Vector column
    :return: Vector column for the result of matrix multiplication
    """
    return _function_factory([matCol, vecCol], _spark_functions().multiplyMatrixVector())


def projectMatrix(matCol, projectionCol):
    """
    Project matrix operation, i.e A * B * A.T where A is the projection matrix

    :param matCol: Matrix column
    :param projectionCol: Projection matrix column
    :return: Matrix column for the result
    """
    return _function_factory([matCol, projectionCol], _spark_functions().projectMatrix())


def randnMultiGaussian(meanArray, covMatrix, seed=0):
    """
    Samples from multivariate gaussian as vector

    :param meanArray: mean of the distribution, either List[Float] or numpy array
    :param covMatrix: covariance of the distribution, either List[List[Float]]] (row major) or numpy 2d array
    :param seed: seed of the rand
    :return: DenseVector column
    """
    root = np.linalg.cholesky(np.array(covMatrix))
    rows, columns = root.shape

    root = arrayToMatrix(
        F.lit(rows), F.lit(columns),
        F.array([F.lit(el) for el in root.reshape(int(rows*columns), order="F").tolist()]))
    mean = arrayToVector(F.array([F.lit(float(el)) for el in meanArray]))

    samples = arrayToVector(F.array([F.randn(seed=seed+el) for el in range(0, len(meanArray))]))
    return _function_factory([mean, root, samples], _spark_functions().scaleToMultiGaussian())